statev2/src/storage/db.rs

```diff
@@ -195,12 +287,177 @@ mod test {
+
+        // Create a second table in the write tx
+        const TABLE_NAME2: &str = "test_table2";
+        let tx = db.new_write_tx().unwrap();
+        tx.create_table(TABLE_NAME2).unwrap();
+
+        tx.write(TABLE_NAME, &key1, &value1).unwrap();
+        tx.write(TABLE_NAME2, &key2, &value2).unwrap();
+        tx.commit().unwrap();
+
+        // Read the values back
+        let tx = db.new_read_tx().unwrap();
+        let v1: Option<TestValue> = tx.read(TABLE_NAME, &key1).unwrap();
+        let v2: Option<TestValue> = tx.read(TABLE_NAME2, &key2).unwrap();
+        tx.commit().unwrap();
+
+        assert_eq!(v1.unwrap(), value1);
+        assert_eq!(v2.unwrap(), value2);
+    }
+
+    /// Tests concurrent transaction isolation
+    #[test]
+    fn test_concurrent_updates() {
+        // Create a mock DB and table
+        let tempdir = tempdir().unwrap();
+        let path = tempdir.path().to_str().unwrap().to_string();
+        let db = Arc::new(DB::new(DbConfig { path: path.clone() }).unwrap());
+
+        db.create_table(TABLE_NAME).unwrap();
+
+        // Write an initial value to the table
+        let key = "counter".to_string();
+        let tx = db.new_write_tx().unwrap();
+        tx.write(TABLE_NAME, &key, &1u64).unwrap();
+        tx.commit().unwrap();
+
+        // Spawn two threads to increment the counter
+        let mut join_handles = Vec::new();
+        for _ in 0..2 {
+            let key_clone = key.clone();
+            let db_clone = db.clone();
+
+            let handle = thread::spawn(move || {
+                let tx = db_clone.new_write_tx().unwrap();
+                let value: u64 = tx.read(TABLE_NAME, &key_clone).unwrap().unwrap();
+                tx.write(TABLE_NAME, &key_clone, &(value + 1)).unwrap();
+                tx.commit().unwrap();
+            });
+
+            join_handles.push(handle);
+        }
+
+        // Await termination
+        join_handles
+            .into_iter()
+            .for_each(|handle| handle.join().unwrap());
+
+        // Now read back the value, it should be incremented twice
+        let tx = db.new_read_tx().unwrap();
+        let value: u64 = tx.read(TABLE_NAME, &key).unwrap().unwrap();
+        tx.commit().unwrap();
+
+        assert_eq!(value, 3);
+    }
+
+    /// Tests recovering from a crash
+    #[test]
+    fn test_crash_recover() {
+        let tempdir = tempdir().unwrap();
+        let path = tempdir.path().to_str().unwrap().to_string();
+
+        // Create a mock DB and table
+        let db = DB::new(DbConfig { path: path.clone() }).unwrap();
+
+        // Set a key
+        let key = "test_key".to_string();
+        let value = TestValue {
+            a: 10,
+            ..Default::default()
+        };
+
+        let tx = db.new_write_tx().unwrap();
+        tx.create_table(TABLE_NAME).unwrap();
+        tx.write(TABLE_NAME, &key, &value).unwrap();
+        tx.commit().unwrap();
+
+        // Drop the db to simulate a crash
+        db.sync().unwrap();
```

Comment:
more out of curiosity, is persistence to disk only ensured by calling `sync`? if there's no periodic syncing going on, then it's on us to make sure we call this often enough / in the right areas, which may cause some annoying heisenbugs

Reply:
No this isn't the only that we write back to disk, it does periodically do so on its own. The `sync` method is just a way to force it